### PR TITLE
Add possibility to capture, edit & replay camera-keyframes

### PR DESCRIPTION
**Pull request by jockekilby**
**_To test:_**

* Create a new sequence, any previously loaded sequence is discarded
```lua
openspace.keyframeRecording.newSequence()
```
* Move camera to a position where you want a keyframe
* Save keyframe, argument is time _within sequence_ where you want the keyframe
```lua
openspace.keyframeRecording.addKeyframe(<time as double>)
```
* Move camera to position of next keyframe
* Save keyframe, argument is time _within sequence_ where you want the keyframe
```lua
openspace.keyframeRecording.addKeyframe(<time as double>)
```
* Keyframes can be removed by calling (index is 0-based)
```lua
openspace.keyframeRecording.removeKeyframe(<index as int>)
```
* Playback sequence, argument is **optional** time _within sequence_ where you want to start playing from. If nothing is specified starts playing from whatever the current time is set to within the sequence
```lua
openspace.keyframeRecording.play(<time as optional double>)
```

**More stuff:_**
Keyframes can be updated by moving the camera and then calling: (argument is 0-based index of keyframe to update)
```lua
openspace.keyframeRecording.updateKeyframe(<index as int>)
```

Keyframes can be moved in time by calling: (arguments are 0-based index of keyframe to update and new time for keyframe)
```lua
openspace.keyframeRecording.moveKeyframe(<index as int>, <time as double>)
```

* A sequence can be saved to disk by calling: 
* Filename can be omitted if the sequence has previously been saved or loaded from disk already
```lua
openspace.keyframeRecording.saveSequence(<filename as optional string>)
```

A sequence can be loaded from disk by calling: (argument is filename **without suffix** so omit _.json_)
```lua
openspace.keyframeRecording.loadSequence(<filename as string>)
```

A playing sequence can be paused:
```lua
openspace.keyframeRecording.pause()
```

You can jump in time within a sequence: (argument is time to jump to)
```lua
openspace.keyframeRecording.setTime(<time as double>)
```

You can query if there is currently a loaded keyframe sequence by:
```lua
openspace.keyframeRecording.hasKeyframeRecording()
```

The currently loaded keyframes can be fetched to the client by calling: 
```lua
openspace.keyframeRecording.getKeyframes()
```

### CODE REVIEW

**Comment by jockekilby** [comment]
Opened as a basis for discussions both about funtionality and implementation details. Please test & comment :)

**Comment by ylvaselling** [comment]
image
I can't build it

**Comment by jockekilby** [comment]
> image
> I can't build it. Configured (clean cache) & generated with Cmake

Huh… 
Had to leave for the weekend but I’ll check it out Monday. 
sorry

**Comment by jockekilby** [comment]
Fixed. Turns out your visual studio version didn't support std::views. I changed into old-fashioned cbegin/cend 😄

**Comment by alexanderbock** [comment]
How about transferring the filename from the `newSequence` function to the `saveSequence` function?  So we'd create a new nameless sequence and when I want to save the keyframe to disk I specify the name?
1. Maybe I want to do a one-off keyframe sequence that I don't intent to save, so why should I give a name?
2. If I want to save multiple versions of the same keyframe sequence just with small changes, it'd be easier.  Instead of: `newSequence("foo")`, ..., `updateKeyframe`, `updateKeyframe`, `saveKeyframe`, rename file on hard disk, `updateKeyframe`, `saveKeyframe` I could do:  `newSequence`, ..., `updateKeyframe`, `updateKeyframe`, `saveKeyframe("foo)`, `updateKeyframe`, `saveKeyframe("bar")` instead

File: src/interaction/keyframerecording.cpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        _keyframes.end(),
+        [sequenceTime](const nlohmann::json& entry) {
+            return sequenceTime < static_cast<double>(entry["timestamp"]["sequence"]);
+        }
+    );
+    _keyframes.insert(it, keyframe);
+}
+
+void KeyframeRecording::updateKeyframe(int index) {
+    nlohmann::json old = _keyframes.at(index);
```
Comment (alexanderbock) [inline src/interaction/keyframerecording.cpp:73]: This will crash if the user provides an `index` that is <0 or bigger than the number of keyframes

File: include/openspace/interaction/keyframerecording.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    void sortKeyframes();
+
+    nlohmann::json newKeyframe(double sequenceTime);
+
+    KeyframeNavigator::CameraPose keyframeToPose(const nlohmann::json& keyframe) const;
+
+    std::vector<nlohmann::json> _keyframes;
+    std::string _filename;
+    bool _isPlaying = false;
+    bool _hasStateChanged = false;
```
Comment (alexanderbock) [inline include/openspace/interaction/keyframerecording.h:78]: I wasn't sure about this default value.  I added it as there was no default value before

File: include/openspace/interaction/keyframerecording.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    static openspace::scripting::LuaLibrary luaLibrary();
+
+private:
+    void sortKeyframes();
+
+    nlohmann::json newKeyframe(double sequenceTime);
+
+    KeyframeNavigator::CameraPose keyframeToPose(const nlohmann::json& keyframe) const;
+
+    std::vector<nlohmann::json> _keyframes;
```
Comment (alexanderbock) [inline include/openspace/interaction/keyframerecording.h:75]: I think it would be nicer to define a `Keyframes` struct that we store in this vector instead.  When we serialize or deserialize, we can convert it into a json object.  That'd make it easier to debug and maintain in the future

File: include/openspace/interaction/keyframerecording.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    std::string _filename;
+    bool _isPlaying = false;
```
Comment (alexanderbock) [inline include/openspace/interaction/keyframerecording.h:77]: The session recording has a similar setting whether we are currently playing back a session recording. I think the UI should query this as well accordingly.  @ylvaselling ?

Comment (ylvaselling) [reply include/openspace/interaction/keyframerecording.h:77]: The UI gets the state of the session recording from the session recording topic, where playback is one of the states. Is this the answer to your question?
https://github.com/OpenSpace/OpenSpace/blob/391c5ef3d1c37e7f1db6ef3789fc52f92825afd5/modules/server/src/topics/sessionrecordingtopic.cpp#L114

Comment (alexanderbock) [reply include/openspace/interaction/keyframerecording.h:77]: Yep 👍 

So the final implementation of this should make use of the same topic

File: src/interaction/keyframerecording.cpp [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+
+namespace {
+    constexpr std::string_view _loggerCat = "KeyframeRecording";
+} // namespace
+
+namespace openspace::interaction {
+
+void to_json(nlohmann::json& j, const KeyframeRecording::Keyframe& keyframe) {
+    j = {
+        { "camera", {
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:50]: You could make these strings as string view constexpr (like in sessionrecording topic) to make it easier to change in the future + prevent spelling mistakes

File: src/interaction/keyframerecording.cpp [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        _keyframes.end(),
+        [sequenceTime](const Keyframe& entry) {
+            return sequenceTime < entry.timestamp.sequenceTime;
+        }
+    );
+    _keyframes.insert(it, keyframe);
+}
+
+void KeyframeRecording::removeKeyframe(int index)
+{
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:127]:
```suggestion
void KeyframeRecording::removeKeyframe(int index) {
```

File: src/interaction/keyframerecording.cpp [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    if (index < 0 || static_cast<size_t>(index) >(_keyframes.size() - 1)) {
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:128]:
```suggestion
    if (index < 0 || static_cast<size_t>(index) > (_keyframes.size() - 1)) {
```

File: src/interaction/keyframerecording.cpp [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+    if (index < 0 || static_cast<size_t>(index) > (_keyframes.size() - 1)) {
+        LERROR(std::format("Index {} out of range", index));
+        return;
+    }
+    Keyframe old = _keyframes[index];
+    _keyframes[index] = newKeyframe(old.timestamp.sequenceTime);
+}
+
+void KeyframeRecording::moveKeyframe(int index, double sequenceTime) {
+    if (index < 0 || static_cast<size_t>(index) >(_keyframes.size() - 1)) {
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:145]:
```suggestion
    if (index < 0 || static_cast<size_t>(index) > (_keyframes.size() - 1)) {
```

File: src/interaction/keyframerecording.cpp [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+        _filename = filename.value();
+    }
+
+    if (_filename.empty()) {
+        LERROR("Failed to save file, reason: Invalid empty file name");
+        return false;
+    }
+
+    nlohmann::json sequence = _keyframes;
+    std::filesystem::path path = absPath("${RECORDINGS}/" + _filename + ".json");
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:166]:
```suggestion
    std::filesystem::path path = absPath(std::format(${RECORDINGS}/{}.json", _filename));
```

File: src/interaction/keyframerecording.cpp [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+    std::ofstream ofs(path);
+    ofs << sequence.dump(2);
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:168]: What does this `2` mean?

File: src/interaction/keyframerecording.cpp [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+    return true;
+}
+
+void KeyframeRecording::loadSequence(std::string filename) {
+    std::filesystem::path path = absPath("${RECORDINGS}/" + filename + ".json");
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:173]: Same as above

File: src/interaction/keyframerecording.cpp [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+    _isPlaying = false;
+}
+
+void KeyframeRecording::setSequenceTime(double sequenceTime) {
+    _sequenceTime = sequenceTime;
+    _hasStateChanged = true;
+}
+
+bool KeyframeRecording::hasKeyframeRecording() const
+{
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:204]:
```suggestion
bool KeyframeRecording::hasKeyframeRecording() const {
```

File: src/interaction/keyframerecording.cpp [hunk 13]
```diff
@@ -0,0 +1,359 @@
...
+    return !_keyframes.empty();
+}
+
+std::vector<ghoul::Dictionary> KeyframeRecording::getKeyframes() const
+{
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:209]:
```suggestion
std::vector<ghoul::Dictionary> KeyframeRecording::getKeyframes() const {
```

File: src/interaction/keyframerecording.cpp [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+    keyframe.camera.scale = scale;
+    keyframe.camera.followFocusNodeRotation = followNodeRotation;
+
+    keyframe.timestamp.application = global::windowDelegate->applicationTime();
+    keyframe.timestamp.sequenceTime = sequenceTime;
+    keyframe.timestamp.simulation = global::timeManager->time().j2000Seconds();
+
+    return keyframe;
+}
+
```
Comment (ylvaselling) [inline src/interaction/keyframerecording.cpp:358]:
```suggestion
```

File: src/interaction/keyframerecording_lua.inl [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+ ****************************************************************************************/
+
+#include <openspace/engine/globals.h>
+#include <optional>
+
+namespace {
+
+// Starts a new sequence of keyframes, any previously loaded sequence is discarded
+[[codegen::luawrap]] void newSequence() {
+
```
Comment (ylvaselling) [inline src/interaction/keyframerecording_lua.inl:32]:
```suggestion
```

File: src/interaction/keyframerecording_lua.inl [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    return openspace::global::keyframeRecording->hasKeyframeRecording();
+}
+
+// Fetches the sequence keyframes as a JSON object
+[[codegen::luawrap]] std::vector<ghoul::Dictionary> getKeyframes() {
+    return openspace::global::keyframeRecording->getKeyframes();
+}
+
+#include "keyframerecording_lua_codegen.cpp"
+
```
Comment (ylvaselling) [inline src/interaction/keyframerecording_lua.inl:98]:
```suggestion
```

File: src/interaction/keyframerecording.cpp (hunk 10 above)
Comment (jockekilby) [reply src/interaction/keyframerecording.cpp:168]: It's _the number of indents for pretty printing_ just like JSON.stringify in JS

**Comment by ylvaselling** [comment]
Wow this is super smooth!
Got an error calling this function
image

**Comment by ylvaselling** [comment]
Hard crash if I try to remove index 0 in an empty recording

image

**Comment by ylvaselling** [comment]
Hard crash trying to play a recording when nothing has been loaded 
image

File: src/interaction/keyframerecording.cpp (hunk 9 above)
Comment (engbergandreas) [reply src/interaction/keyframerecording.cpp:166]: Can't be done because "${RECORDINGS} is not compile time constant :'(

File: src/interaction/keyframerecording_lua.inl [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+ * Returns `true` if there currently is a sequence loaded, otherwise `false`
+ */
+[[codegen::luawrap]] bool hasKeyframeRecording() {
+    return openspace::global::keyframeRecording->hasKeyframeRecording();
+}
+
+/**
+ * Fetches the sequence keyframes as a JSON object
+ */
+ [[codegen::luawrap]] std::vector<ghoul::Dictionary> keyframes() {
```
Comment (WeirdRubberDuck) [inline src/interaction/keyframerecording_lua.inl:172]:
```suggestion
[[codegen::luawrap]] std::vector<ghoul::Dictionary> keyframes() {
```

File: src/interaction/keyframerecording_lua.inl [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+[[codegen::luawrap]] void loadSequence(std::string filename) {
+    openspace::global::keyframeRecording->loadSequence(std::move(filename));
+}
+
+/**
+ * Playback sequence optionally from the specified `sequenceTime` or if not specified
+ *  starts playing from the beginning
+ *
+ * \param sequenceTime The time in seconds at which to start playing the sequence, or
+ * beginning if omitted
```
[the remark this answers is not in the document]
Comment (WeirdRubberDuck) [reply src/interaction/keyframerecording_lua.inl:126]: Maybe rephrase last half sa a new sentence: 

"If omitted, the playback starts at the beginning of the sequence. "

File: src/interaction/keyframerecording.cpp [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+    for (const nlohmann::json& keyframeJson : jsonKeyframes) {
+        Keyframe keyframe = keyframeJson;
+        _keyframes.push_back(keyframe);
+    }
+    _filename = filename;
+}
+
+void KeyframeRecording::play() {
+    ghoul_assert(hasKeyframeRecording(), "Keyframe sequence can't be empty")
```
Comment (WeirdRubberDuck) [inline src/interaction/keyframerecording.cpp:236]:
```suggestion
    ghoul_assert(hasKeyframeRecording(), "Keyframe sequence can't be empty");
```

File: include/openspace/interaction/keyframerecording.h [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+    void loadSequence(std::string filename);
+    void preSynchronization(double dt);
+    void play();
+    void pause();
+    void setSequenceTime(double sequenceTime);
+    bool hasKeyframeRecording() const;
+    std::vector<ghoul::Dictionary> keyframes() const;
+
+    /**
+     * \return The Lua library that contains all Lua functions available to affect the
```
Comment (WeirdRubberDuck) [inline include/openspace/interaction/keyframerecording.h:67]: This comment seems outdated ?

File: src/interaction/keyframerecording.cpp [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+    constexpr const char* X = "x";
+    constexpr const char* Y = "y";
+    constexpr const char* Z = "z";
+    constexpr const char* W = "w";
+}
+
+namespace openspace::interaction {
+
+void to_json(nlohmann::json& j, const KeyframeRecording::Keyframe& keyframe) {
+    j = {
```
Comment (WeirdRubberDuck) [inline src/interaction/keyframerecording.cpp:68]: How about splitting this one up into multiple smaller json objects and then combining them at the end? 

It current

File: src/interaction/keyframerecording.cpp [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+void KeyframeRecording::setSequenceTime(double sequenceTime) {
+    ghoul_assert(sequenceTime >= 0, "Sequence time must be positive");
+
+    _sequenceTime = sequenceTime;
+    _hasStateChanged = true;
+    LINFO(std::format("Set sequence time to {}", sequenceTime));
+}
+
+void KeyframeRecording::jumpToKeyframe(int index)
+{
```
Comment (WeirdRubberDuck) [inline src/interaction/keyframerecording.cpp:260]:
```suggestion
void KeyframeRecording::jumpToKeyframe(int index) {
```

File: src/interaction/keyframerecording_lua.inl [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+    if (index < 0) {
+        throw ghoul::lua::LuaError("Index value must be positive");
+    }
+    openspace::global::keyframeRecording->removeKeyframe(index);
+}
+
+/**
+ * Update the camera position of a keyframe at the specified index.
+ *
+ * \param index The 0-based index of e keyframe to update
```
Comment (WeirdRubberDuck) [inline src/interaction/keyframerecording_lua.inl:68]:
```suggestion
 * \param index The 0-based index of the keyframe to update
```

File: include/openspace/interaction/keyframerecording.h (hunk 20 above)
Comment (WeirdRubberDuck) [reply include/openspace/interaction/keyframerecording.h:68]: Remove it? Not really needed
